pkg/didcomm/protocol/didexchange/states.go

```diff
@@ -428,52 +431,87 @@ func (ctx *context) getDestinationFromDID(id string) (*service.Destination, erro
 		return nil, err
 	}
 
-	pubKeys, err := getPublicKeys(didDoc, supportedPublicKeyType)
+	return prepareDestination(didDoc)
+}
+
+func getRecipientKeys(didDoc *did.Doc) ([]string, error) {
+	didCommService, err := getDidCommService(didDoc)
 	if err != nil {
 		return nil, err
 	}
 
-	recepientKey := string(pubKeys[0].Value)
+	if len(didCommService.RecipientKeys) == 0 {
+		return nil, fmt.Errorf("missing recipient keys in did-communication service")
+	}
 
-	serviceEndpoint, err := getServiceEndpoint(didDoc)
-	if err != nil {
-		return nil, err
+	var recipientKeys []string
+
+	for _, keyID := range didCommService.RecipientKeys {
+		key, err := getPublicKey(keyID, didDoc)
+		if err != nil {
+			return nil, err
+		}
+
+		if isSupportedKeyType(key.Type) {
+			recipientKeys = append(recipientKeys, string(key.Value))
+		}
 	}
 
-	return &service.Destination{
-		RecipientKeys:   []string{recepientKey},
-		ServiceEndpoint: serviceEndpoint,
-		RoutingKeys:     []string{recepientKey},
-	}, nil
+	if len(recipientKeys) == 0 {
+		return nil, fmt.Errorf("recipient keys in did-communication service not supported")
+	}
+
+	return recipientKeys, nil
 }
 
-func getServiceEndpoint(didDoc *did.Doc) (string, error) {
-	for _, s := range didDoc.Service {
-		if s.Type == serviceType {
-			return s.ServiceEndpoint, nil
+func getPublicKey(id string, didDoc *did.Doc) (*did.PublicKey, error) {
+	for _, key := range didDoc.PublicKey {
+		if key.ID == id {
+			return &key, nil
 		}
 	}
 
-	return "", errors.New("service not found in DID document")
+	return nil, fmt.Errorf("key not found in DID document: %s", id)
+}
+
+func isSupportedKeyType(keyType string) bool {
+	return keyType == ed25519KeyType
 }
 
-func prepareDestination(didDoc *did.Doc) *service.Destination {
-	var srvEndPoint string
-	for _, v := range didDoc.Service {
-		srvEndPoint = v.ServiceEndpoint
+func getDidCommService(didDoc *did.Doc) (*did.Service, error) {
+	const notFound = -1
+	index := notFound
+
+	for i, s := range didDoc.Service {
+		if s.Type == didCommServiceType {
+			if index == notFound || didDoc.Service[index].Priority > s.Priority {
+				index = i
+			}
+		}
+	}
+
+	if index == notFound {
+		return nil, fmt.Errorf("service not found in DID document: %s", didCommServiceType)
```

Comment:
Is error the correct behaviour if the service endpoint isn't defined in the DID document?

Reply:
How do we pick recipient keys and sender verification key then?

Reply:
I double-checked with @llorllale : As per did-comm convention did document has to have "did-communication" service.